src/tools/similarity.ts

```diff
@@ -59,39 +75,62 @@ export class SimilarityTool<TProviderOptions> extends Tool<
     this.register();
   }
 
-  protected async _run(
-    input: ToolInput<this>,
-    options?: SimilarityToolRunOptions<TProviderOptions>,
-  ) {
-    const { query, documents } = input;
+  wrapTool<T extends Tool, S extends ZodSchema>(tool: T, schema: S) {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
 
-    const results = await this.options.provider(
-      {
-        query,
-        documents,
-      },
-      options?.provider,
-    );
-
-    const resultsWithDocumentIndices = results.map(({ score }, idx) => ({
-      documentIndex: idx,
-      score,
-    }));
-    const sortedResultsWithDocumentIndices = R.sortBy(resultsWithDocumentIndices, [
-      ({ score }) => score,
-      "desc",
-    ]);
-    const filteredResultsWithDocumentIndices = sortedResultsWithDocumentIndices.slice(
-      0,
-      options?.maxResults ?? this.options.maxResults,
-    );
+    return (config: {
+      toTool: (input: z.output<S>) => ToolInput<T>;
+      fromTool: (
+        input: z.output<S>,
```

Comment:
Given that SimilarityTool is also a tool, it may be clearer to name these `toWrappedTool` and `fromWrappedTool`